File: src/main/java/kaalka/KaalkaFile.java
package kaalka;

import java.util.*;
import java.nio.charset.StandardCharsets;

/**
 * KaalkaFile: Chunked file encryption/decryption
 */
public class KaalkaFile {
    public static List<KaalkaEnvelope> encryptFileChunks(byte[] fileBytes, String senderId, String receiverId, String timestamp) {
        List<KaalkaEnvelope> chunks = new ArrayList<>();
        int chunkSize = 1024; // 1KB chunks
        int totalChunks = (fileBytes.length + chunkSize - 1) / chunkSize;
        for (int i = 0; i < totalChunks; i++) {
            int start = i * chunkSize;
            int end = Math.min(start + chunkSize, fileBytes.length);
            byte[] chunk = Arrays.copyOfRange(fileBytes, start, end);
            String chunkTime = timestamp + ":" + i;
            KaalkaEnvelope env = KaalkaProtocol.encryptEnvelope(new String(chunk, StandardCharsets.ISO_8859_1), senderId, receiverId, chunkTime, 120, i + 1);
            chunks.add(env);
        }
        return chunks;
    }

    public static byte[] decryptFileChunks(List<KaalkaEnvelope> encryptedChunks, String receiverId, String timestamp) throws Exception {
        List<byte[]> outChunks = new ArrayList<>();
        for (int i = 0; i < encryptedChunks.size(); i++) {
            KaalkaEnvelope env = encryptedChunks.get(i);
            String chunkTime = timestamp + ":" + i;
            String pt = new KaalkaProtocol().decryptEnvelope(env, receiverId, chunkTime);
            outChunks.add(pt.getBytes(StandardCharsets.ISO_8859_1));
        }
        int totalLen = outChunks.stream().mapToInt(b -> b.length).sum();
        byte[] fileBytes = new byte[totalLen];
        int pos = 0;
        for (byte[] chunk : outChunks) {
            System.arraycopy(chunk, 0, fileBytes, pos, chunk.length);
            pos += chunk.length;
        }
        return fileBytes;
    }
}

File: src/main/kotlin/kaalka/KaalkaFile.kt
package kaalka

/**
 * Chunked file encryption/decryption
 */
object KaalkaFile {
    fun encryptFileChunks(
        fileBytes: ByteArray,
        senderId: String,
        receiverId: String,
        timestamp: String? = null
    ): List<KaalkaEnvelope> {
        val chunkSize = 1024
        val totalChunks = (fileBytes.size + chunkSize - 1) / chunkSize
        return (0 until totalChunks).map { i ->
            val start = i * chunkSize
            val end = minOf(start + chunkSize, fileBytes.size)
            val chunk = fileBytes.sliceArray(start until end)
            val chunkTime = (timestamp ?: KaalkaUtils.utcNow()) + ":$i"
            KaalkaProtocol.encryptEnvelope(chunk.toString(Charsets.ISO_8859_1), senderId, receiverId, chunkTime, 120, i + 1)
        }
    }

    fun decryptFileChunks(
        chunks: List<KaalkaEnvelope>,
        receiverId: String,
        timestamp: String? = null
    ): ByteArray {
        val outChunks = chunks.mapIndexed { i, env ->
            val chunkTime = (timestamp ?: env.timestamp) + ":$i"
            KaalkaProtocol.decryptEnvelope(env, receiverId, chunkTime).toByteArray(Charsets.ISO_8859_1)
        }
        val totalLen = outChunks.sumOf { it.size }
        val fileBytes = ByteArray(totalLen)
        var pos = 0
        for (chunk in outChunks) {
            System.arraycopy(chunk, 0, fileBytes, pos, chunk.size)
            pos += chunk.size
        }
        return fileBytes
    }
}
